### PR TITLE
LB-633: Add "Listening Activity" statistics (API)

### PR DESCRIPTION
**Pull request by ishaanshah**
# Description
This PR aims to add an API endpoint for getting users listening activity.

# Note
This should be merged after #919.

### CODE REVIEW

File: listenbrainz/webserver/views/stats_api.py [hunk 1]
```diff
@@ -345,6 +346,86 @@ def get_recording(user_name):
     }})
 
 
+@stats_api_bp.route("/user/<user_name>/listening_activity")
```
Comment (paramsingh) [inline listenbrainz/webserver/views/stats_api.py:349]:
```suggestion
@stats_api_bp.route("/user/<user_name>/listening-activity")
```

We generally use dashes in API endpoints

File: listenbrainz/webserver/views/stats_api.py [hunk 2]
```diff
@@ -345,6 +346,86 @@ def get_recording(user_name):
...
+                "user_id": "ishaanshah"
+            }
+        }
+    .. note::
+        - This endpoint is currently in beta
+        - The example above shows the data for three days only, however we calculate the statistics for
+          the current time range and the previous time range. For example for weekly statistics the data
+          is calculated for the current as well as the past week.
+
+    :param range: Optional, time interval for which statistics should be collected, possible values are ``week``,
```
Comment (paramsingh) [inline listenbrainz/webserver/views/stats_api.py:392]:
```suggestion
    :param range: Optional, time interval for which statistics should be returned, possible values are ``week``,
```

File: listenbrainz/webserver/views/stats_api.py [hunk 3]
```diff
@@ -345,6 +346,86 @@ def get_recording(user_name):
...
+        raise APINotFound("Cannot find user: {}".format(user_name))
+
+    stats_range = request.args.get('range', default='all_time')
+    if not _is_valid_range(stats_range):
+        raise APIBadRequest("Invalid range: {}".format(stats_range))
+
+    try:
+        stats = db_stats.get_user_listening_activity(user['id'], stats_range)
+    except ValidationError:
+        # If the stored stats are not correct, we should should not return them
```
Comment (paramsingh) [inline listenbrainz/webserver/views/stats_api.py:413]:
```suggestion
        # If the stored stats are not correct, we should not return them
```

File: listenbrainz/tests/integration/test_stats_api.py [hunk 4]
```diff
@@ -705,3 +712,108 @@ def test_entity_range_stat_not_calculated(self):
...
+        self.assert200(response)
+        data = json.loads(response.data)['payload']
+
+        sent_from = self.listening_activity_payload['from_ts']
+        received_from = data['from_ts']
+        self.assertEqual(sent_from, received_from)
+        sent_to = self.listening_activity_payload['to_ts']
+        received_to = data['to_ts']
+        self.assertEqual(sent_to, received_to)
+        sent_listening_activty = self.listening_activity_payload['listening_activity']
```
Comment (paramsingh) [inline listenbrainz/tests/integration/test_stats_api.py:728]: Spelling of activity here and in other places.

File: listenbrainz/tests/integration/test_stats_api.py [hunk 5]
```diff
@@ -705,3 +712,108 @@ def test_entity_range_stat_not_calculated(self):
...
+        received_listening_activity = data['listening_activity']
+        self.assertListEqual(sent_listening_activty, received_listening_activity)
+        self.assertEqual(data['user_id'], self.user['musicbrainz_id'])
+
+    def test_artist_stat_all_time(self):
```
Comment (paramsingh) [inline listenbrainz/tests/integration/test_stats_api.py:733]: This test name needs to be listening activity

Comment (paramsingh) [reply listenbrainz/tests/integration/test_stats_api.py:733]: And other tests too.

File: listenbrainz/tests/integration/test_stats_api.py (hunk 4 above)
Comment (ishaanshah) [reply listenbrainz/tests/integration/test_stats_api.py]: Fixed

File: listenbrainz/tests/integration/test_stats_api.py (hunk 5 above)
Comment (ishaanshah) [reply listenbrainz/tests/integration/test_stats_api.py]: Fixed

**Pull request merged**